### DISCUSSION TITLE
Add equality function into `atom`

### DISCUSSION THREAD

**Discussion by daprahamian** [pmndrs/jotai#2001] [Ideas]
I find myself doing a lot of:

```ts
const myAtom = selectAtom(atom(...), v => v, (a, b) => {
  // custom equality function here
});
```

Any thoughts on adding the equality function as part of the atom definition itself? something like one of the following?

```ts
// As part of atom constructor
const myAtom = atom(
  (get) => ...,
  (get, set, update) => ...,
  (a, b) => /* custom equality function */
);

// As an add on similar to `onMount`
const myAtom = atom(...);
myAtom.isEqual = (a, b) => {
  // custom equality function here
}
```

**Comment by dai-shi** [comment]
Theoretically, there's no notion of "previous" values in pure function model, and depending on it is against typical atom structure. `selectAtom` is an **escape hatch** to allow extra optimization in an impure way. It's sometimes useful, but shouldn't be overused.



> ```ts
> const myAtom = selectAtom(atom(...), v => v, (a, b) => {
>   // custom equality function here
> });
> ```

That just looks good to me. If you feel repetitive, you can create a custom util function on your end.

**Reply by daprahamian** [reply]
Thanks for the reply @dai-shi! I understand that from the theoretical model, there is no concept of a previous value. However, from the perspective of recomputing dependencies, previous values are definitely used. You can see it being used in the store [here](https://github.com/pmndrs/jotai/blob/main/src/vanilla/store.ts#LL532C32-L532C48).

That said, I can get by with a util function on my end, so I'm fine leaving it as is. Again, thank you!

**Reply by dai-shi** [reply]
just a note: the `write` part of atom is imperative so previous values can be used (not only in library code internally, but also in your application code). the `read` is declarative and pure.

```js
const anAtom = atom(
  (get) => ...,
  (get, set, arg) => {
     const prevValue = get(anotherAtom);
     set(anotherAtom, ...);
  }
);
```
